Docs/feature_memory_protection.md

```diff
@@ -1,359 +1,210 @@
-[NX](#no-executeexecute-neverexecute-disable-attribute-nxxnxd) via the [NXCOMPAT](#nxcompat) DLL flag, the logic will
-cease to set the [NX](#no-executeexecute-neverexecute-disable-attribute-nxxnxd) attribute on allocations of memory
-this type. Using the BlockImagesWithoutNxFlag in the [Image Protection Policy](image-protection-policy) will prevent
-images which don't support [NXCOMPAT](#nxcompat).
-- EfiLoaderData
-- EfiBootServicesCode - Same note as EfiLoaderCode.
-- EfiBootServicesData
-- EfiRuntimeServicesCode - Same note as EfiLoaderCode.
-- EfiRuntimeServicesData
-- EfiConventionalMemory
-- EfiUnusableMemory
-- EfiACPIReclaimMemory
-- EfiACPIMemoryNVS
-- EfiMemoryMappedIO
-- EfiMemoryMappedIOPortSpace
-- EfiPalCode
-- EfiPersistentMemory
-- OEMReserved
-- OSReserved
-
-## **Page Guards**
-
-### Summary
-
-The HeapGuardPageType policy implements guard pages on the specified memory types to detect heap overflow. If a bit
-is set, a guard page will be added before and after the
-corresponding type of page allocated if there's enough free pages for all of them. Guard pages are set to NOT PRESENT so any attempt
-to access them will cause a [#PF](#page-fault-exception-pf). The system will do its best to ensure that only one guard page separates two allocated pages to avoid wasted space.
-
-### Overhead
-**O(n)** time where n is the number of page allocations/frees. Because there are 2 extra pages allocated for
-every call to AllocatePages(), **O(n)** space is also required.
-
-### Available Settings for DXE and MM
-
-- EfiReservedMemoryType
-- EfiLoaderCode
-- EfiLoaderData
-- EfiBootServicesCode
-- EfiBootServicesData
-- EfiRuntimeServicesCode
-- EfiRuntimeServicesData
-- EfiConventionalMemory
-- EfiUnusableMemory
-- EfiACPIReclaimMemory
-- EfiACPIMemoryNVS
-- EfiMemoryMappedIO
-- EfiMemoryMappedIOPortSpace
-- EfiPalCode
-- EfiPersistentMemory
-- OEMReserved
-- OSReserved
-
-## **Pool Guards**
-
-### Summary
+- **UefiNullDetection**: Enable NULL pointer detection for DXE.
+- **DisableEndOfDxe**: Disable NULL pointer detection just after the end of DXE protocol
+is installed.
+- **DisableReadyToBoot**: Disable NULL pointer detection just after the ready to boot
+protocol is installed.
+
+The **MM environment** only has a single option indicating whether NULL detection
+is active or not.
+
+## FreeMemoryReadProtected
+
+If enabled, all EfiConventionalMemory (free memory) will be marked with the
+[EFI_MEMORY_RP](#efi_memory_rp) attribute. This can be used in conjunction with the
+NX setting for EfiConventionalMemory.
+
+## Image Protection Policy
+
+This policy enables a loaded EFI image to have [EFI_MEMORY_XP](#efi_memory_xp) to its
+DATA sections and [EFI_MEMORY_RO](#efi_memory_ro) to its CODE sections. Loaded EFI
+images must adhere to the following rules for this policy to work:
+
+1. The PE code section and data sections are not merged.
+2. The PE image sections must be page aligned.

```

Comment:
Can this be updated to say that image sections of DXE_RUNTIME_DRVIERs need their image sections RUNTIME_PAGE_ALLOCATION_GRANULARITY aligned. Probably a note that the file alignment does not need to be RUNTIME_PAGE_ALLOCATION_GRANULARITY aligned but that the loaded in memory alignment needs to be RUNTIME_PAGE_ALLOCATION_GRANULARITY aligned.

Reply:
Added to the list of rules a point which describes the requirements for runtime drivers